saddle-core/src/main/scala/org/saddle/scalar/ScalarTagByte.scala

```diff
@@ -23,4 +23,5 @@ object ScalarTagByte extends ScalarTagAny[Byte] {
   override def parse(s: String) = s.toByte
   override def makeSorter(implicit ord: ORD[Byte]): Sorter[Byte] =
     Sorter.byteSorter
+  override def missing: Byte = Byte.MinValue
```

Comment:
Unfortunately this is binary incompatible, at least MiMa flags it. :( 
However this form is the more correct one: this is specialized while the previous one is erased to Object.

Reply:
Uh, I didn't expect this one to cause a binary incompat. I'll check further into this. Thanks for flagging it.

Reply:
Here is the corresponding mima error:
```[error]  * static method missing()java.lang.Object in class org.saddle.scalar.ScalarTagByte has a different result type in current version, where it is Byte rather than java.lang.Object
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.saddle.scalar.ScalarTagByte.missing")
```
I'm wondering whether this isn't a mima false positive. My assumption is that `missing()Byte` is turned into `missing()java.lang.Object` after erasure and will accept boxed primitives because it's an override of `missing()T`.

Reply:
I checked with javap and I think it is the other way around: the override specializes it. So that is actually better, except not compatible.

Reply:
Here is a minimized repro: 
```scala
trait Default[T] {
    def default: T = null.asInstanceOf[T]
}

object DefaultInt extends Default[Int] {
    override def default: Int = 0
}
```
These class files are produced, Default.class, DefaultInt.class and DefaultInt$.class. Here is javap output: 
```java
// Default.class
Compiled from "Default.scala"
public interface Default<T> {
  public static java.lang.Object default$(Default);
  public T default();
  public static void $init$(Default);
}

// DefaultInt.class
Compiled from "Tag.scala"
public final class DefaultInt {
  public static int default();
}

// DefaultInt$.class
Compiled from "Tag.scala"
public final class DefaultInt$ implements Default<java.lang.Object> {
  public static final DefaultInt$ MODULE$;
  public static {};
  public int default();
  public java.lang.Object default();
}
```
Here is mima's error
```
[error]  * static method default()java.lang.Object in class DefaultInt has a different result type in current version, where it is Int rather than java.lang.Object
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("DefaultInt.default")
```

It seems scala produces specialized variants, one in the object DefaultInt$.class  and one static on the associated class DefaultInt.class. An boxed variant seems present `public java.lang.Object default();` in the object. Would mima be looking at `public in default();` and fail instead of looking further?

Reply:
Ah the mima message refers to the static method in the associated class. It's not about the object methods. Is that generated for java interop or are these methods called in scala?

Reply:
A scala Int can come in the form of boxed or unboxed so I believe this is the reason that both kinds are emitted.

But here is the type signature of the method in the java bytecode for the variant with override (the one you pasted):
```
public final class DefaultInt
{
  public static int default();
    descriptor: ()I
    flags: ACC_PUBLIC, ACC_STATIC
    Code:
      stack=1, locals=0, args_size=0
         0: getstatic     #16                 // Field DefaultInt$.MODULE$:LDefaultInt$;
         3: invokevirtual #18                 // Method DefaultInt$.default:()I
         6: ireturn
}
```

and without the override (just `object DefaultInt extends Default[Int]{}`)

```
{
  public static java.lang.Object default();
    descriptor: ()Ljava/lang/Object;
    flags: ACC_PUBLIC, ACC_STATIC
    Code:
      stack=1, locals=0, args_size=0
         0: getstatic     #16                 // Field DefaultInt$.MODULE$:LDefaultInt$;
         3: invokevirtual #18                 // Method DefaultInt$.default:()Ljava/lang/Object;
         6: areturn
}
```

One returns a boxed int (a java.lang.Object) the other returns a primitive.
.

Reply:
That is a static method added for java interop, but potentially never used from scala bytecode. In this example: 
```scala
object DefaultIntTest {
    DefaultInt.default
}
```
javap output: 
```
Compiled from "Default.scala"
public final class DefaultIntTest$ {
  public static final DefaultIntTest$ MODULE$;

  public static {};
    Code:
       0: new           #2                  // class DefaultIntTest$
       3: dup
       4: invokespecial #12                 // Method "<init>":()V
       7: putstatic     #14                 // Field MODULE$:LDefaultIntTest$;
      10: getstatic     #19                 // Field DefaultInt$.MODULE$:LDefaultInt$;
      13: invokevirtual #23                 // Method DefaultInt$.default:()I  --> object method, not class static method
      16: pop
      17: return
}
```
At the bytecode level, `DefaultInt.default` calls the object method, and not the associated class static method. So, my assumption here is that this is binary compatible for scala programs, not for java ones. I expect there is no java users of saddle, so we could potentially ignore these compat errors.

Reply:
I posted a question on the scala user forum about this: https://users.scala-lang.org/t/binary-compatibility-primitive-types-and-java-static-methods/8314

Reply:
```
trait Default[T] {
    def default: T = null.asInstanceOf[T]
}

object DefaultInt extends Default[Int] {
 //   override def default: Int = 0
}

object DefaultIntTest {
   def x : Int =  DefaultInt.default
}
```

That `x` in the above has a box/unbox trip:

```
  public int x();
    descriptor: ()I
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: getstatic     #19                 // Field DefaultInt$.MODULE$:LDefaultInt$;
         3: invokevirtual #23                 // Method DefaultInt$.default:()Ljava/lang/Object;
         6: invokestatic  #29                 // Method scala/runtime/BoxesRunTime.unboxToInt:(Ljava/lang/Object;)I
         9: ireturn
```

While if the comment is removed:

```
trait Default[T] {
    def default: T = null.asInstanceOf[T]
}

object DefaultInt extends Default[Int] {
    override def default: Int = 0
}

object DefaultIntTest {
   def x : Int =  DefaultInt.default
}
```

then the box is gone:
```
  public int x();
    descriptor: ()I
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: getstatic     #19                 // Field DefaultInt$.MODULE$:LDefaultInt$;
         3: invokevirtual #22                 // Method DefaultInt$.default:()I
         6: ireturn
```

Reply:
Another observation: against the base version (not this branch's one) the java static methods aren't likely to be used from java:
```java
package org.saddle;
import org.saddle.scalar.ScalarTagByte;

public class BinCompatJava {
    public static void test() {
        // these two don't compile
        // incompatible types: java.lang.Object cannot be converted to java.lang.Byte
        // incompatible types: java.lang.Object cannot be converted to byte
        Byte b = ScalarTagByte.missing();
        byte b = ScalarTagByte.missing();
        // this compiles fine
        Object o = ScalarTagByte.missing();
    }
}

Reply:
Interesting. Thanks for the investigation. I need some time to digest this, I will write back in a couple of days, possible after the weekend.

Yes we don't pursue a java api here.